django_mongodb/query.py

```diff
@@ -0,0 +1,372 @@
+
+        rhs, rhs_params = child.process_rhs(self.compiler, self.connection)
+
+        lookup_type = child.lookup_name
+
+        # Since NoSql databases generally don't support aggregation or
+        # annotation, pass true unless the query has a get_aggregation() method.
+        # It's a little troubling however that the _nomalize_lookup_value
+        # method seems to only use this value in the case that the value is an
+        # iterable and the lookup_type equals isnull.
+        annotation = (
+            self.get_aggregation(using=self.connection)[None]
+            if hasattr(self, "get_aggregation")
+            else True
+        )
+        value = rhs_params
+        packed = child.lhs.get_group_by_cols()[0]
+        alias = packed.alias
+        column = packed.target.column
+        field = child.lhs.output_field
+        opts = self.query.model._meta
+        if alias and alias != opts.db_table:
+            raise DatabaseError("This database doesn't support JOINs and multi-table inheritance.")
+
+        # For parent.child_set queries the field held by the constraint
+        # is the parent's primary key, while the field the filter
+        # should consider is the child's foreign key field.
+        if column != field.column:
+            if not field.primary_key:
+                raise DatabaseError(
+                    "This database doesn't support filtering on non-primary key ForeignKey fields."
+                )
+
+            field = next(f for f in opts.fields if f.column == column)
+
+        value = self._normalize_lookup_value(lookup_type, value, field, annotation)
+
+        return field, lookup_type, value
+
+    def _normalize_lookup_value(self, lookup_type, value, field, annotation):
+        """
+        Undo preparations done by Field.get_db_prep_lookup() not suitable for
+        MongoDB, and pass the lookup argument through
+        DatabaseOperations.prep_lookup_value().
+        """
+        # Undo Field.get_db_prep_lookup() putting most values in a list
+        # (a subclass may override this, so check if it's a list) and
+        # losing the (True / False) argument to the "isnull" lookup.
+        if lookup_type not in ("in", "range", "year") and isinstance(value, tuple | list):
+            if len(value) > 1:
+                raise DatabaseError(
+                    "Filter lookup type was %s; expected the filter argument "
+                    "not to be a list. Only 'in'-filters can be used with "
+                    "lists." % lookup_type
+                )
+            value = annotation if lookup_type == "isnull" else value[0]
+
+        # Remove percent signs added by Field.get_db_prep_lookup() for LIKE
+        # queries.
+        if lookup_type in ("startswith", "istartswith"):
+            value = value[:-1]
+        elif lookup_type in ("endswith", "iendswith"):
+            value = value[1:]
+        elif lookup_type in ("contains", "icontains"):
+            value = value[1:-1]
```

Comment:
This should probably still check if there's a percent sign there. That or could we subclass `Field` to remove this layer?

Reply:
While it wouldn't be much extra computation to be defensive, I can't imagine a way it could break, especially in a way that wouldn't be caught by the test suite.

The comment is actually out of date. The percent signs are now added by [PatternLookup.process_rhs()](https://github.com/django/django/blob/e72049aa6302411d8cdf2e798e15fa38e76b92fc/django/db/models/lookups.py#L586). I don't think we can monkey patch that without breaking usage of this backend alongside a SQL backend (and the same comment applies if the logic was in `Field.get_db_prep_lookup()`).